### PR TITLE
Fix vex export returning invalid CycloneDX

### PR DESCRIPTION
**Pull request by SaberStrat**
### Description

Fixes exported VEX and BOMs with vulnerabilities not conforming to CycloneDX 1.5.

### Addressed Issue

#3834 

### Additional Details

The general idea is to deduplicate vulnerabilities based on all `Vulnerability` fields.

I settled on an implementation with one HashSet that tracks all `Vulnerability` objects seen when converting `Finding`s into `Vulnerability`s.

It requires an enhancement of the cyclonedx-core-java library, namely adding custom `equals` and `hashCode` methods to the `Vulnerability` class for fieldwise comparison: https://github.com/CycloneDX/cyclonedx-core-java/issues/463.

Other possible implementations I considered: 
1. Comparison of `Vulnerability`s based on all of its fields _except_ for `affects` and accumulate affects objects in one vulnerability if all other `Vulnerability` fields have otherwise equal values. This came from an interpretation of a remark from the original issue reporter in the issue report. But the CycloneDX schema doesn't really require an accumulation, if I'm not mistaken. 
I considered a few different implementations of this (in case an accumulation was not wrong after all).
    1. a HashMap of `Vulnerability`s to HashSets of `Affect`s. This was the way I'd have gone with
    2. generating the affects list in its own convert method. I think this would've meant introducing two more methods. Could do this, if an accumulation of affects is a possibility after all
    3. appending to an `affects` list defined outside of and passed to `Vulnerability`'s convert() calls. Was not sure about having `convert()` convert vulnerability objects AND modify a state of a list passed to it.

### Checklist

- [x] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- [x] This PR fixes a defect, and I have provided tests to verify that the fix is effective
- [ ] ~This PR implements an enhancement, and I have provided tests to verify that it works as intended~
- [ ] ~This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)~
- [ ] ~This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly~

### CODE REVIEW

File: src/test/java/org/dependencytrack/parser/cyclonedx/CycloneDXExporterTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import org.dependencytrack.tasks.scanners.AnalyzerIdentity;
+
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThatNoException;
+
+public class CycloneDXExporterTest extends PersistenceCapableTest {
+
+    @Test // https://github.com/DependencyTrack/dependency-track/issues/3834
+    public void testSchemaValidityOfExportedVexWithMultipleComponentsWithSameVulnerability() throws GeneratorException {
```
Comment (nscuro) [inline src/test/java/org/dependencytrack/parser/cyclonedx/CycloneDXExporterTest.java:17]: Consider moving this to [`VexResourceTest`](https://github.com/DependencyTrack/dependency-track/blob/master/src/test/java/org/dependencytrack/resources/v1/VexResourceTest.java), and asserting the entire response JSON.

Also, a test case that demonstrates what happens when `componentA` and `componentB` have different analyses (i.e. one is suppressed the other is not, or one has sate `IN_TRIAGE` the other `EXPLOITABLE`) would be good.

If we de-duplicate vulnerabilities without the `Analysis` in mind, we might drop important data.

**Review by nscuro — CHANGES_REQUESTED** [review]

File: src/test/java/org/dependencytrack/parser/cyclonedx/CycloneDXExporterTest.java (hunk 1 above)
Comment (SaberStrat) [reply src/test/java/org/dependencytrack/parser/cyclonedx/CycloneDXExporterTest.java:17]: Yeah I'm re-implementing it such that the whole Vulnerability object is used for comparison. Still in the process of making it work.

Will do what you wrote with the tests.

Comment (SaberStrat) [reply src/test/java/org/dependencytrack/parser/cyclonedx/CycloneDXExporterTest.java]: Done

**Comment by nscuro** [comment]
@SaberStrat cyclonedx-core-java 9.0.5 was [released earlier today](https://github.com/CycloneDX/cyclonedx-core-java/releases/tag/cyclonedx-core-java-9.0.5), which includes your changes. You should be able to bump the library version now to make your implementation work.

**Comment by SaberStrat** [comment]
cyclonedx-core-java bumped and rebased to master.

**Comment by SaberStrat** [comment]
@nscuro I take it adding escaped quotes to the expected error messages in the test methods is not a proper solution for this?

**Comment by nscuro** [comment]
Yup! Nothing special to do on our side since we merely forward the error messages. But still good to have such tests to be made aware of *when* these things change.

**Comment by SaberStrat** [comment]
Rinse and repeat. This time, I ran the full test suite locally and it passed.

Should I cleanup/squash the commits down logically?

**Comment by nscuro** [comment]
> Should I cleanup/squash the commits down logically?

Not necessary.

**Review by nscuro — APPROVED** [review]
Thanks @SaberStrat! I'll also backport this to v4.11.6.

**Pull request merged**